src/com/palmergames/bukkit/towny/command/TownCommand.java

```diff
@@ -3524,6 +3537,90 @@ public static void parseTownUnclaimCommand(Player player, String[] split) {
+			return;
+		}
+
+		if (TownySettings.getMaxResidentsPerTown() > 0 && (remainingTown.getResidents().size() + town.getResidents().size()) > TownySettings.getMaxResidentsPerTown()) {
+			TownyMessaging.sendErrorMsg(player, Translation.of("msg_town_merge_err_too_many_residents", TownySettings.getMaxResidentsPerTown()));
+			return;
+		}
+
+		if ((remainingTown.getTownBlocks().size() + town.getTownBlocks().size()) > TownySettings.getMaxTownBlocks(remainingTown)) {
+			TownyMessaging.sendErrorMsg(player, Translation.of("msg_town_merge_err_too_many_townblocks", TownySettings.getMaxTownBlocks(remainingTown)));
+			return;
+		}
+
+		if (!BukkitTools.isOnline(town.getMayor().getName()) || town.getMayor().isNPC()) {
+			TownyMessaging.sendErrorMsg(player, Translation.of("msg_town_merge_other_offline", town.getName(), town.getMayor().getName()));
+			return;
+		}
+
+		double cost = 0;
+		if (TownyEconomyHandler.isActive()) {
+			try {
+				cost = (double) remainingTown.getTownBlockCostN(town.getTownBlocks().size())/2; // Cost is cost to claim the other's townblocks divided by 2.
+				System.out.println(cost);
+				if (!remainingTown.getAccount().canPayFromHoldings(cost)) {
+					TownyMessaging.sendErrorMsg(player, Translation.of("msg_town_merge_err_not_enough_money", remainingTown.getAccount().getHoldingBalance(), cost));
+				}
+			} catch (TownyException | EconomyException e) {
+				TownyMessaging.sendErrorMsg(player, Translation.of("msg_town_merge_failed"));
+				return;
+			}			
+		}
+
+		// Send town merge request.
+		TownyMessaging.sendMsg(resident, Translation.of("msg_town_merge_request_sent", town.getName()));
+		TownyMessaging.sendMsg(town.getMayor(), Translation.of("msg_town_merge_request_received", remainingTown.getName(), resident.getName(), remainingTown.getName()));
+
+		final Town finalTown = town;
+		final Town finalRemainingTown = remainingTown;
+		final double finalCost = cost;
+		Confirmation.runOnAccept(() -> {
+			if (TownyEconomyHandler.isActive()) {
+				// Check if town can still pay merging costs.
+				try {
+					if (!finalRemainingTown.getAccount().canPayFromHoldings(finalCost))
+						TownyMessaging.sendErrorMsg(player, Translation.of("msg_town_merge_err_not_enough_money", finalRemainingTown.getAccount().getHoldingBalance(), finalCost));
+				} catch (EconomyException ignored) {}
+			}
+
+			TownPreMergeEvent townPreMergeEvent = new TownPreMergeEvent(finalTown, finalRemainingTown);
+			Bukkit.getPluginManager().callEvent(townPreMergeEvent);
+			if (townPreMergeEvent.isCancelled()) {
+				TownyMessaging.sendErrorMsg(town.getMayor().getPlayer(), townPreMergeEvent.getCancelMessage());
+				TownyMessaging.sendErrorMsg(player, townPreMergeEvent.getCancelMessage());
+				return;
+			}
+
+			// Start merge
+			TownyUniverse.getInstance().getDataSource().mergeTown(finalRemainingTown, finalTown);
+
+			TownMergeEvent townMergeEvent = new TownMergeEvent(finalTown, finalRemainingTown);

```

Comment:
This is after the Town is removed so the finalTown will never exist at this point. It would be useful to give the name and UUID of the removed Town so things can be cleaned up in the plugin using this event.